### PR TITLE
Add `torch.vdot`

### LINKED ISSUE #42747: Add torch.vdot similar to numpy.vdot to calculate the complex dot product
For two complex numbers the [dot product](https://en.wikipedia.org/wiki/Dot_product#Complex_vectors) is defined as:

Screen Shot 2020-08-07 at 3 57 51 PM

The `torch.vdot(a, b)` function (similar to [numpy.vdot](https://numpy.org/doc/stable/reference/generated/numpy.vdot.html))will treat complex numbers differently than dot(a, b). If the first argument is complex the complex conjugate of the first argument will be used for the calculation of the dot product.

cc @ezyang @anjali411 @dylanbespalko @mruberry @rgommers

### PR DESCRIPTION
**Pull request by kshitij12345**
Fixes #42747

### CODE REVIEW

File: aten/src/ATen/native/Blas.cpp
Comment (mruberry) [inline aten/src/ATen/native/Blas.cpp:152]: `at::empty(0, ...`

This actually trips up experienced PyTorch developers all the time. If we take a look at `empty`:

https://github.com/pytorch/pytorch/blob/eb47940c0a773ec52e67778dbe01918f6b162271/aten/src/ATen/native/TensorFactories.cpp#L102

the size comes from here:

https://github.com/pytorch/pytorch/blob/eb47940c0a773ec52e67778dbe01918f6b162271/aten/src/ATen/native/TensorFactories.cpp#L121

which is computed using

https://github.com/pytorch/pytorch/blob/c9a1fc2b31ded472c63b227ad24859bc14131f42/aten/src/ATen/Utils.h#L96

So if we run the following program:

```
std::cout << prod_intlist({}) << std::endl;
  std::cout << prod_intlist(0) << std::endl;
```

it will print 1, 0. It is kinda confusing that an empty initializer list produces a tensor with one element and putting a zero in produces a tensor with no elements, but I try to remember it by recalling that zero specifies the size.

Comment (mruberry) [reply aten/src/ATen/native/Blas.cpp:152]: Oh wait. You're immediately populating the value into result, not resizing it (which I assumed since it's such a common pattern). You're doing everything right.

Comment (mruberry) [reply aten/src/ATen/native/Blas.cpp:152]: (Although what about `empty(1, ...` for readability?)

File: aten/src/ATen/native/BlasKernel.cpp
Comment (mruberry) [inline aten/src/ATen/native/BlasKernel.cpp:227]: This is really elegant.

File: test/test_torch.py
Comment (mruberry) [inline test/test_torch.py:16662]: To fix the XLA issues I think you can flip this check:

```
if self.device_type == 'cpu':
  ...
else:
  y_np = y.cpu().numpy()
```

File: test/test_torch.py
Comment (mruberry) [inline test/test_torch.py:16667]: Instead of setting the atol/rtol directly I think you can just the @precisionOverride decorator. To make the tests pass I think you'll need:

```
@precisionOverride({torch.cfloat: 1e-4})
```

File: test/test_torch.py
Comment (mruberry) [inline test/test_torch.py:16681]: See above comment - this can probably be removed

File: torch/_torch_docs.py
Comment (mruberry) [inline torch/_torch_docs.py:2167]: "vdot(input, other)" to be consistent with other binary PyTorch functions

File: torch/_torch_docs.py
Comment (mruberry) [inline torch/_torch_docs.py:2181]: Instead of randomly generating these complex tensors, how about some simple examples that a reader could compute themselves if they wanted? Like (1 +2j, 3 - 1j) or something?

File: aten/src/ATen/native/native_functions.yaml
Comment (mruberry) [inline aten/src/ATen/native/native_functions.yaml:1156]: "Tensor self, Tensor other" to be consistent with other binary PyTorch functions

File: torch/_torch_docs.py
Comment (mruberry) [inline torch/_torch_docs.py:2165]: Update https://github.com/pytorch/pytorch/blob/master/docs/source/torch.rst, too

File: aten/src/ATen/native/native_functions.yaml
Comment (ngimel) [reply aten/src/ATen/native/native_functions.yaml:1156]: `dot` just above is using `Tensor self, Tensor tensor`. Also, what's the reasoning behind having `dot` as `function, method`, and `vdot` as just `function`?

Comment (mruberry) [reply aten/src/ATen/native/native_functions.yaml:1156]: Ug. I guess that's OK although `dot`'s signature is also wrong.

Methods are nice (and maybe still required to be put into method_tests?).

**Review by mruberry — CHANGES_REQUESTED** [review]
Overall looks really good as usual, @kshitij12345!

What about adding a derivative, like we have for dot:

https://github.com/pytorch/pytorch/blob/eb47940c0a773ec52e67778dbe01918f6b162271/tools/autograd/derivatives.yaml#L395

And updating method_tests(): 

https://github.com/pytorch/pytorch/blob/eb47940c0a773ec52e67778dbe01918f6b162271/torch/testing/_internal/common_methods_invocations.py#L632

to test it? 

@anjali411 for help with the derivative.

File: aten/src/ATen/native/native_functions.yaml
Comment (mruberry) [inline aten/src/ATen/native/native_functions.yaml:1163]: Better put vdot after all the dot functions (and add vdot.out?)

File: test/test_torch.py
Comment (mruberry) [inline test/test_torch.py:16684]: You don't need to reconstruct v1. You can create 

```
v3 = torch.randn(1, dtype=dtype, device=device).expand(numel)
compare_with_numpy_bin_op(torch.vdot, np.vdot, v1, v3, relaxed_tolerance)
compare_with_numpy_bin_op(torch.vdot, np.vdot, v3, v1, relaxed_tolerance)
```

File: test/test_torch.py
Comment (mruberry) [inline test/test_torch.py:16660]: Can this unify test_dot and test_vdot by creating two tests that call a common helper?

```
_test_dot_vdot_helper(self, device, dtype, op, ref):
   ...

test_dot(...):
  ...

test_vdot(...):
  ...
```

Where both tests call `_test_dot_vdot_helper`. This would also test our current dot implementation vs. NumPy's dot, which would be nice, and get it testing on CUDA, which would also be good.

Comment (mruberry) [inline test/test_torch.py:16660]: Would you also add tests for:

-  dot and vdot getting arguments of different dtypes, incorrect number of dimensions, and mismatched devices (this test is a little weird to write in the device generic framework, but you can check that your current device is a cuda device and then create a tensor on that device + a cpu tensor to do it)
- out behavior, including mismatched out dtype
- after the method variant is added, you can add a test here, too:

https://github.com/pytorch/pytorch/blob/eb47940c0a773ec52e67778dbe01918f6b162271/test/test_torch.py#L19786

File: torch/_torch_docs.py
Comment (mruberry) [inline torch/_torch_docs.py:2169]: What about something like:

"Computes the dot product (inner product) of input the complex conjugate of other."

And not having to explain in a note how vdot is distinct from dot?

File: aten/src/ATen/native/cuda/LinearAlgebra.cu
Comment (anjali411) [inline aten/src/ATen/native/cuda/LinearAlgebra.cu:423]:
```
>>> np.vdot(np.array([2]), np.array([2+3j]))
(4+6j)
>>> np.vdot(np.array([2+3j]), np.array([2]))
(4-6j)
```
maybe it's worth adding a comment that we only call `dot_cuda` when `self` is not complex because we want the above mentioned behavior.

Comment (anjali411) [reply aten/src/ATen/native/cuda/LinearAlgebra.cu:423]: nvm I guess we throw an error for input tensors of different dtype:
```
>>> torch.dot(torch.tensor([1j]), torch.tensor([2]))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: dot : expected both vectors to have same dtype, but found ComplexFloat and Long
```
 I think we should add type promotion for `dot` and `vdot` to better align with numpy.

cc. @mruberry

File: aten/src/ATen/native/native_functions.yaml
Comment (anjali411) [inline aten/src/ATen/native/native_functions.yaml:1158]: `function, method`

File: test/test_torch.py
Comment (anjali411) [reply test/test_torch.py:16660]: https://github.com/pytorch/pytorch/pull/43004/files#r470634289

File: torch/_torch_docs.py
Comment (anjali411) [reply torch/_torch_docs.py:2169]: I think we should include the note in the description because it brings more clarity to the description.

```
"Computes the dot product (inner product) of two tensors. The vdot(a, b) function handles complex numbers differently than dot(a, b). If the first argument is complex the complex conjugate of the first argument is used for the calculation of the dot product."
```

**Review by anjali411 — CHANGES_REQUESTED** [review]
I left some minor comments. LGTM overall :) 
1. We should add a not implemented definition in `derivatives.yaml` until the JAX vs tf issue is resolved.

```
- name: dot(Tensor self, Tensor other) -> Tensor
  self: 'not_implemented("vdot: self")'
  other: 'not_implemented("vdot: other")'
```

2. We should not throw an error for input tensors of different dtypes to be more consistent with numpy. cc. @mruberry 
There will be some merge conflicts once this PR is merged https://github.com/pytorch/pytorch/pull/42745

File: aten/src/ATen/native/cuda/LinearAlgebra.cu
Comment (mruberry) [reply aten/src/ATen/native/cuda/LinearAlgebra.cu:423]: Type promotion would be nice but I think it's OK if it's not in this first PR. Type promotion is a little tricky to implement today when not using TensorIterator.

File: aten/src/ATen/native/native_functions.yaml
Comment (kshitij12345) [reply aten/src/ATen/native/native_functions.yaml:1156]: For this operator I avoided `method` for 2 reasons.

* https://numpy.org/doc/stable/reference/generated/numpy.vdot.html Signature of numpy doesn't have it.
* Second since it returns a 1-element tensor, using it in method chaining might lead to confusion. (Maybe overly cautious)

Do let me know if that makes or we should indeed have method variant.

Comment (ngimel) [reply aten/src/ATen/native/native_functions.yaml:1156]: numpy does not have methods for most things, but pytorch does. `dot` is a method, so it would make sense for `vdot` to be a method too.

File: aten/src/ATen/native/cuda/LinearAlgebra.cu
Comment (kshitij12345) [reply aten/src/ATen/native/cuda/LinearAlgebra.cu:423]: Yeah. I was thinking about it. But since the plan is to use existing `dot` for real types, wasn't sure as to how to go about having type promotion.

Also other place where it diverges from numpy is that numpy operator supports broadcasting while `dot` doesn't. We can add broadcasting logic before passing the inputs to `dot`. What do you feel about it?

Comment (mruberry) [reply aten/src/ATen/native/cuda/LinearAlgebra.cu:423]: I would file follow-up issues for broadcasting + type promotion. We may even want to add architecture to support doing these things easily outside of TensorIterator.

File: aten/src/ATen/native/native_functions.yaml
Comment (kshitij12345) [reply aten/src/ATen/native/native_functions.yaml:1163]: Would out variant be required here? As output really is a single element tensor. Also numpy doesn't have `out` in its [signature](https://numpy.org/doc/stable/reference/generated/numpy.vdot.html)

However as mentioned `dot` does have a out variant.

Comment (mruberry) [reply aten/src/ATen/native/native_functions.yaml:1163]: It's very easy to add and we should be consistent with `torch.dot`.

File: test/test_torch.py
Comment (kshitij12345) [reply test/test_torch.py:16660]: Makes sense. Was thinking of doing the same but since `dot` has an `out` variant the testing would have slightly diverged so did not touch `dot`'s tests.

File: aten/src/ATen/native/native_functions.yaml
Comment (kshitij12345) [reply aten/src/ATen/native/native_functions.yaml:1163]: Sure. Will do that.

File: aten/src/ATen/native/native_functions.yaml
Comment (kshitij12345) [reply aten/src/ATen/native/native_functions.yaml:1156]: Sure.

File: aten/src/ATen/native/cuda/LinearAlgebra.cu
Comment (ngimel) [reply aten/src/ATen/native/cuda/LinearAlgebra.cu:423]: numpy `vdot` does not support broadcasting, and we don't try to follow numpy `dot` behavior:
```
In [5]: np.vdot([2,3], [3])                                                                                                                                                                                                         
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-5-6192cfa76e6f> in <module>
----> 1 np.vdot([2,3], [3])

<__array_function__ internals> in vdot(*args, **kwargs)

ValueError: cannot reshape array of size 1 into shape (2,)
```

Comment (anjali411) [reply aten/src/ATen/native/cuda/LinearAlgebra.cu:423]: @mruberry yeah that sounds good. out of curiosity -- what's tricky about implementing type promotion when not using TensorIterator? 

@kshitij12345  
```
if (!self.is_complex()) {
    return dot_cuda(self, other);
}
``` will give us the desired behavior for (real, complex) dot product when the type promotion is enabled, so we should still add a note documenting [that](https://github.com/pytorch/pytorch/pull/43004#discussion_r470631911). This is so that when the type promotion is enabled in future, we have the behavior already documented.

Comment (mruberry) [reply aten/src/ATen/native/cuda/LinearAlgebra.cu:423]: Type promotion challenges:

- sometimes have to be careful to preserve your inputs (luckily not the case here)
- compute the result type
- cast inputs to the result type
- validate safe casting to out
- copy to out (if necessary)
- write a custom test for your op's type promotion behavior

It's not the end of the world. In this case we would want to change the behavior of dot, too, so it seems separable.

File: aten/src/ATen/native/native_functions.yaml
Comment (kshitij12345) [reply aten/src/ATen/native/native_functions.yaml]: Done.

File: test/test_torch.py
Comment (kshitij12345) [reply test/test_torch.py]: Done.

File: test/test_torch.py
Comment (kshitij12345) [reply test/test_torch.py]: Done. Thanks!

**Comment by kshitij12345** [comment]
Should I add `onlyCPUandCUDA` for the failing test below?

https://app.circleci.com/pipelines/github/pytorch/pytorch/200838/workflows/c40940ee-dd25-4d72-88fa-64692058d03d/jobs/6735701

Also I see that the `test_vdot_vs_numpy_cpu_complex128/64` is failing on ROCM-CPU (surprising as others are passing).
@anjali411 Did you see similar thing for #42745 
https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-xenial-rocm3.5.1-py3.6-test2/4352/console

Any idea what the reason might me?

Thanks!

**Comment by anjali411** [comment]
> Should I add `onlyCPUandCUDA` for the failing test below?
> 
> https://app.circleci.com/pipelines/github/pytorch/pytorch/200838/workflows/c40940ee-dd25-4d72-88fa-64692058d03d/jobs/6735701
> 
> Also I see that the `test_vdot_vs_numpy_cpu_complex128/64` is failing on ROCM-CPU (surprising as others are passing).
> @anjali411 Did you see similar thing for #42745
> https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-xenial-rocm3.5.1-py3.6-test2/4352/console
> 
> Any idea what the reason might me?
> 
> Thanks!

yeah the rocm failures are probably because `zdotc_` and `cdotc_` symbols from openblas install are getting linked and called instead of the static inline versions that wrap the cblas calls. Adding `cdotc_` and `zdotc_` in the else condition like [here](https://github.com/pytorch/pytorch/pull/42745/files#diff-217694a98df68fe93858b39ac4a30f39R20) would likely resolve the issue.

**Comment by kshitij12345** [comment]
> yeah the rocm failures are probably because zdotc_ and cdotc_ symbols from openblas install are getting linked and called instead of the static inline versions that wrap the cblas calls. Adding cdotc_ and zdotc_ in the else condition like here would likely resolve the issue.

@anjali411 Thank you! Will try that.

File: aten/src/ATen/cuda/CUDABlas.cpp
Comment (anjali411) [inline aten/src/ATen/cuda/CUDABlas.cpp:551]: can we just add a cuda to hip mapping for `cuComplex` -> `rocblas_float_complex` and `cuDoubleComplex` -> `rocblas_double_complex`?

cc. @jeffdaily

**Review by anjali411 — APPROVED** [review]
Good to go once the comments about documentation and inline comments are resolved.

File: test/test_torch.py
Comment (ngimel) [inline test/test_torch.py:16731]: can you please remove `test_dot`? You are covering all tests from test_dot here.

File: test/test_torch.py
Comment (ngimel) [inline test/test_torch.py:16748]: can you generate v1 and/or v2 with strides more than 1?

File: test/test_torch.py
Comment (ngimel) [inline test/test_torch.py:16804]: You should be able to test `dot` for complex too, the PR has landed I think

Comment (anjali411) [reply test/test_torch.py:16804]: yeah

File: aten/src/ATen/cuda/CUDABlas.cpp
Comment (jeffdaily) [reply aten/src/ATen/cuda/CUDABlas.cpp:551]: Let's try it.  I didn't see cuComplex used in any other context besides blas.

File: test/test_torch.py
Comment (kshitij12345) [reply test/test_torch.py:16786]: Done. Thanks!

File: test/test_torch.py
Comment (kshitij12345) [reply test/test_torch.py:16722]: Have added a case with strides more than 1. Thanks!

File: test/test_torch.py
Comment (kshitij12345) [reply test/test_torch.py:16705]: Done.

File: aten/src/ATen/native/LinearAlgebra.cpp
Comment (mruberry) [inline aten/src/ATen/native/LinearAlgebra.cpp:560]: Use:

https://github.com/pytorch/pytorch/blob/dae94ed02299cd1e9840ce6e49ea142b4d118b2d/aten/src/ATen/native/Resize.cpp#L9

to resize outputs

File: tools/autograd/derivatives.yaml
Comment (mruberry) [inline tools/autograd/derivatives.yaml:407]: Does this mean that vdot, even when it defers to dot, is also not differentiable, or will that work?

File: aten/src/ATen/native/LinearAlgebra.cpp
Comment (mruberry) [inline aten/src/ATen/native/LinearAlgebra.cpp:559]: `... self, const Tensor& other)`

File: torch/_torch_docs.py
Comment (mruberry) [inline torch/_torch_docs.py:2344]: `..., other, *, out)`

File: torch/_torch_docs.py
Comment (mruberry) [inline torch/_torch_docs.py:2353]: Something like:

Args:
  input (Tensor): first tensor in the dot product. Its conjugate is used if it's complex.
  other (Tensor): second tensor in the dot product.

Keyword args:
  {out}

File: torch/_torch_docs.py
Comment (mruberry) [inline torch/_torch_docs.py:2364]: If you do use a construction like {out} then you'll need to add a `.format(` here.

File: aten/src/ATen/native/LinearAlgebra.cpp
Comment (kshitij12345) [reply aten/src/ATen/native/LinearAlgebra.cpp:560]: Adding it to `vdot_out`.

Should I also use it for `dot_out`?

https://github.com/pytorch/pytorch/blob/fa6b34b54c731938327c8e30e08b287a10b86b0a/aten/src/ATen/native/LinearAlgebra.cpp#L552-L557

Seems like a small change and the tests are already there for out variant.

File: tools/autograd/derivatives.yaml
Comment (kshitij12345) [reply tools/autograd/derivatives.yaml:407]: Yes `vdot` wouldn't be differentiable even when it will defer to `dot`, which I feel is better than `vdot` being differentiable for real types and not for complex.

File: torch/_torch_docs.py
Comment (kshitij12345) [reply torch/_torch_docs.py:2356]: Thanks!!!
 Done.

File: aten/src/ATen/cuda/CUDABlas.cpp
Comment (anjali411) [reply aten/src/ATen/cuda/CUDABlas.cpp:551]: @kshitij12345 can you try add a mapping for `cuComplex` -> `rocblas_float_complex` in `torch/utils/hipify/cuda_to_hip_mappings.py`?

Comment (kshitij12345) [reply aten/src/ATen/cuda/CUDABlas.cpp:550]: Have tried. 

Added the following (however not sure if it is correct)
```
("cuComplex", ("rocblas_float_complex", CONV_TYPE, API_RUNTIME)),
("cuDoubleComplex", ("rocblas_double_complex", CONV_TYPE, API_RUNTIME)),
```

Verify with CI as I don't have the machinery to do ROCm build locally.

Comment (jeffdaily) [reply aten/src/ATen/cuda/CUDABlas.cpp:550]: LGTM.  CI build succeeded, but test jobs got assigned to a CI host we only recently found to be broken.  That host was just taken offline.  Let's try a pytorchbot retest to verify.

Comment (anjali411) [reply aten/src/ATen/cuda/CUDABlas.cpp:550]: I think it should be:
```
("cuComplex", ("rocblas_float_complex", CONV_TYPE, API_BLAS)),
("cuDoubleComplex", ("rocblas_double_complex", CONV_TYPE, API_BLAS)),
```

Comment (jeffdaily) [reply aten/src/ATen/cuda/CUDABlas.cpp:550]: That's fair, yes, perhaps to avoid issues in the future if we actually do any major filtering of these meta data tags like `API_BLAS`.  The only tags we actually check for in the hipify_python.py script are `API_CAFFE2` and `API_PYTORCH` to differentiate between substitutions that are specific to one or other framework.  Otherwise, we apply the given substitution as-is to all sources.

File: aten/src/ATen/native/LinearAlgebra.cpp
Comment (mruberry) [reply aten/src/ATen/native/LinearAlgebra.cpp]: Please do! Thank you.

File: tools/autograd/derivatives.yaml
Comment (mruberry) [reply tools/autograd/derivatives.yaml:407]: Sounds good. Just wanted to be sure we're clear on the behavior. Thanks for being so thorough, @kshitij12345.

File: aten/src/ATen/cuda/CUDABlas.cpp
Comment (kshitij12345) [reply aten/src/ATen/cuda/CUDABlas.cpp:550]: Sure. Will do that.

However `cuComplex` is part of `cuComplex.h`, so should there be a API_COMPLEX or something similar going forward? Just curious to know more as I don't know much about the ROCm effort.

Thanks!

**Comment by kshitij12345** [comment]
Gentle ping:)

**Comment by jeffdaily** [comment]
ROCm CI passed with latest changes.  LGTM.

File: torch/_torch_docs.py
Comment (mruberry) [reply torch/_torch_docs.py:2346]: May still need to inline this note per @anjali411's feedback. I'll let her have final say over doc string.

**Comment by mruberry** [comment]
You also need to update tensors.rst like torch.rst.

**Comment by kshitij12345** [comment]
> You also need to update tensors.rst like torch.rst.

Thanks had missed that. Have also updated the description of the function.

File: torch/_torch_docs.py
Comment (anjali411) [reply torch/_torch_docs.py:2366]: @kshitij12345 nit -- can you change the (new line) formatting to:
```
"Computes the dot product (inner product) of two tensors. The vdot(a, b) function handles complex numbers differently than dot(a, b). If the first argument is complex the complex conjugate of the first argument is used for the calculation of the dot product."
```

Comment (kshitij12345) [reply torch/_torch_docs.py]: Done. Thanks!

**Comment by kshitij12345** [comment]
Gentle Ping :)

**Review by mruberry — APPROVED** [review]
Ensuring Phabricator diff is stamped.

**Comment by anjali411** [comment]
@kshitij12345 thanks for the reminder! the FB tests for this PR were failing, hence it took a while. can you rebase the PR?

**Pull request closed without merging**
